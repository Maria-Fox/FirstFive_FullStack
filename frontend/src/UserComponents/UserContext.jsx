import React from "react";

// Create user context instance to later provide in App.js and use throughout application. Holds auth user

const UserContext = React.createContext();

export default UserContext;