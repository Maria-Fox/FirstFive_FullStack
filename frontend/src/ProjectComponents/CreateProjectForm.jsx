import React, { useState, useContext } from "react";
import { useNavigate } from "react-router-dom";
import API from "../API";
import UserContext from "../UserComponents/UserContext";
import AlertNotification from "../Common/AlertNotifications";
import { Form, FormGroup, Button, CardBody, Card, Label, Input, CardText } from "reactstrap"

const CreateProjectForm = () => {

  // ***************************************************************
  let navigate = useNavigate();
  let { authUser, matchedProjectIds, setMatchedProjectIds } = useContext(UserContext);

  // Set as undefined so user cannot submit an empty project proposal.
  let initial_state = {
    name: "",
    project_desc: "",
    timeframe: "",
    github_repo: ""
  };

  let [projData, setProjData] = useState(initial_state);
  let [errors, setErrors] = useState(null);

  // ***************************************************************

  let handleChange = (e) => {
    let { name, value } = e.target;
    setProjData(projData => ({
      ...projData,
      [name]: value
    }));
  };

  // ***************************************************************

  async function handleSubmit(e) {
    e.preventDefault();
    try {
      if(projData.name.length < 1 || projData.project_desc.length < 1){
        setErrors("Please add a proejct name and description.");
        return;
      }
      let response = await API.createProject({ owner_username: authUser, ...projData });
      // user is instantly matched with project. They can find it under the 'posts' Nav item.

      let id = response.id;
      // Newly created proj id goes into matches.
      setMatchedProjectIds([...matchedProjectIds, id]);
      navigate(`/projects/created/by/${authUser}`);

    } catch (e) {
      // if()
      setErrors(e);
      return;
    };
  };

  // ***************************************************************


  return (
    <>
      <h1 className="text-center text-white pt-2 mt-2">Create a Project</h1>

      <Card className="container">
        {errors ? <AlertNotification messages={errors} /> : null}

        <CardText className="p-3">Important: You are automatically "matched" with any project you create. Then, as the project owner you may choose whether or not you want to be a project member. If you simply want to suggest a project to the community you can later transfer ownership to another interested user.  </CardText>

        <CardBody>
          <Form onSubmit={handleSubmit}>

            <FormGroup>
              <Label htmlFor="name" >Project Name
                <Input
                  type="text"
                  id="name"
                  value={projData.name}
                  name="name"
                  required
                  placeholder="Workout App"
                  onChange={handleChange}
                >
                </Input>
              </Label>
            </FormGroup>

            <FormGroup>
              <Label htmlFor="project_desc" >Project Description
                <Input
                  type="textarea"
                  id="project_desc"
                  value={projData.project_desc}
                  name="project_desc"
                  required
                  cols="35"
                  rows="8"
                  placeholder="Create a full-stack application where users can track their workouts!"
                  onChange={handleChange}
                >
                </Input>
              </Label>
            </FormGroup>


            <FormGroup>
              <Label htmlFor="timeframe" >Timeframe
                <Input
                  type="text"
                  id="timeframe"
                  value={projData.timeframe}
                  name="timeframe"
                  placeholder="3 weeks"
                  required
                  onChange={handleChange}
                >
                </Input>
              </Label>
            </FormGroup>

            <FormGroup>
              <Label htmlFor="github_repo" >Github Repo
                <Input
                  type="text"
                  id="github_repo"
                  value={projData.github_repo}
                  name="github_repo"
                  placeholder="https://github.com/"
                  onChange={handleChange}
                >
                </Input>
              </Label>
            </FormGroup>

            <Button> Submit</Button>

          </Form>
        </CardBody>
      </Card>
    </>

  )
}

export default CreateProjectForm;