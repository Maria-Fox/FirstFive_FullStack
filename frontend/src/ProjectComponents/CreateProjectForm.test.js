import UserContext from "../UserComponents/UserContext";
import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import CreateProjectForm from "./CreateProjectForm";

describe("Render CreateProjectForm", function () {


  test("Smoke- Valid user renders create project form",
    function () {

      let authUserData = { username: "softwareDev1", matchedProjectIds: [] };

      render(
        <MemoryRouter>
          <UserContext.Provider value={authUserData} >
            <CreateProjectForm />
          </UserContext.Provider>
        </MemoryRouter>
      )
    });

  test("Snapshot- Valid user renders create project form",
    function () {

      let authUserData = { username: "softwareDev1", matchedProjectIds: [] };

      const { asFragment } = render(
        <MemoryRouter>
          <UserContext.Provider value={authUserData} >
            <CreateProjectForm />
          </UserContext.Provider>
        </MemoryRouter>
      );

      expect(asFragment()).toMatchSnapshot();
    });


})