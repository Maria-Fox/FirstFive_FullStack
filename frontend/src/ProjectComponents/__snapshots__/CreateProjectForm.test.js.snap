// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render CreateProjectForm Snapshot- Valid user renders create project form 1`] = `
<DocumentFragment>
  <h1
    class="text-center text-white pt-2 mt-2"
  >
    Create a Project
  </h1>
  <div
    class="container card"
  >
    <p
      class="p-3 card-text"
    >
      Important: You are automatically "matched" with any project you create. Then, as the project owner you may choose whether or not you want to be a project member. If you simply want to suggest a project to the community you can later transfer ownership to another interested user.  
    </p>
    <div
      class="card-body"
    >
      <form
        class=""
      >
        <div
          class="mb-3"
        >
          <label
            class="form-label"
            for="name"
          >
            Project Name
            <input
              class="form-control"
              id="name"
              name="name"
              placeholder="Workout App"
              required=""
              type="text"
              value=""
            />
          </label>
        </div>
        <div
          class="mb-3"
        >
          <label
            class="form-label"
            for="project_desc"
          >
            Project Description
            <textarea
              class="form-control"
              cols="35"
              id="project_desc"
              name="project_desc"
              placeholder="Create a full-stack application where users can track their workouts!"
              required=""
              rows="8"
            />
          </label>
        </div>
        <div
          class="mb-3"
        >
          <label
            class="form-label"
            for="timeframe"
          >
            Timeframe
            <input
              class="form-control"
              id="timeframe"
              name="timeframe"
              placeholder="3 weeks"
              required=""
              type="text"
              value=""
            />
          </label>
        </div>
        <div
          class="mb-3"
        >
          <label
            class="form-label"
            for="github_repo"
          >
            Github Repo
            <input
              class="form-control"
              id="github_repo"
              name="github_repo"
              placeholder="https://github.com/"
              type="text"
              value=""
            />
          </label>
        </div>
        <button
          class="btn btn-secondary"
        >
           Submit
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
